rust2vec/src/vocab.rs

```diff
@@ -17,42 +17,94 @@ pub enum WordIndex {
     Subword(Vec<usize>),
 }
 
-#[derive(Debug, Eq, PartialEq)]
-pub enum Vocab {
-    /// Vocabulary that only consists of words.
-    SimpleVocab {
-        indices: HashMap<String, usize>,
-        words: Vec<String>,
-    },
-    SubwordVocab {
-        indices: HashMap<String, usize>,
-        words: Vec<String>,
-        min_n: u32,
-        max_n: u32,
-        buckets_exp: u32,
-    },
+/// Vocabulary without subword units.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct SimpleVocab {
+    indices: HashMap<String, usize>,
+    words: Vec<String>,
 }
 
-impl Vocab {
-    const BOW: char = '<';
-    const EOW: char = '>';
-
-    pub fn new_simple_vocab(words: impl Into<Vec<String>>) -> Self {
+impl SimpleVocab {
+    pub fn new(words: impl Into<Vec<String>>) -> Self {
         let words = words.into();
-        let indices = Self::create_indices(&words);
-        Vocab::SimpleVocab { words, indices }
+        let indices = create_indices(&words);
+        SimpleVocab { words, indices }
+    }
+}
+
+impl ReadChunk for SimpleVocab {
+    fn read_chunk<R>(read: &mut R) -> Result<Self, Error>
+    where
+        R: Read + Seek,
+    {
+        let chunk_id = ChunkIdentifier::try_from(read.read_u32::<LittleEndian>()?)
+            .ok_or(err_msg("Unknown chunk identifier"))?;
+        ensure!(
+            chunk_id == ChunkIdentifier::SimpleVocab,
+            "Cannot read chunk {:?} as SimpleVocab",
+            chunk_id
+        );
+
+        // Read and discard chunk length.
+        read.read_u64::<LittleEndian>()?;
+
+        let vocab_len = read.read_u64::<LittleEndian>()? as usize;
+        let mut words = Vec::with_capacity(vocab_len);
+        for _ in 0..vocab_len {
+            let word_len = read.read_u32::<LittleEndian>()? as usize;
+            let mut bytes = vec![0; word_len];
+            read.read_exact(&mut bytes)?;
+            let word = String::from_utf8(bytes)?;
+            words.push(word);
+        }
+
+        Ok(SimpleVocab::new(words))
     }
+}
+
+impl WriteChunk for SimpleVocab {
+    fn chunk_identifier(&self) -> ChunkIdentifier {
+        ChunkIdentifier::SimpleVocab
+    }
+
+    fn write_chunk<W>(&self, write: &mut W) -> Result<(), Error>
+    where
+        W: Write + Seek,
+    {
+        let chunk_len = self.words.iter().map(|w| w.len() as u64 + 4).sum::<u64>() + 8;
```

Comment:
`+ 4` because `word.len()` is serialized as `u32`?
`+ 8` is somewhat unclear to me but I'd assume it's because `self.words.len()` ends up as 8 bytes?
Short comment might be nice to follow the logic

Reply:
I have added comment, but I'll make a separate PR replacing this with size_of_val.